chain/index/events.go

```diff
@@ -0,0 +1,548 @@
+		tipsetKeyCid = f.TipsetCid.Bytes()
+	case f.MinHeight >= 0 && f.MinHeight == f.MaxHeight:
+		tipsetKeyCid, err = si.getTipsetKeyCidByHeight(ctx, f.MinHeight)
+		if err != nil {
+			if err == ErrNotFound {
+				// this means that this is a null round and there exist no events for this epoch
+				return nil
+			}
+
+			return xerrors.Errorf("failed to get tipset key cid by height: %w", err)
+		}
+	default:
+		// Filter doesn't specify a specific tipset
+		return nil
+	}
+
+	// If we couldn't determine a specific tipset, return ErrNotFound
+	if tipsetKeyCid == nil {
+		return ErrNotFound
+	}
+
+	// Check if the determined tipset is indexed
+	exists, err := si.isTipsetIndexed(ctx, tipsetKeyCid)
+	if err != nil {
+		return xerrors.Errorf("failed to check if tipset is indexed: %w", err)
+	}
+
+	if exists {
+		return nil // Tipset is indexed
+	}
+
+	return ErrNotFound // Tipset is not indexed
+}
+
+// getTipsetKeyCidByHeight retrieves the tipset key CID for a given height.
+func (si *SqliteIndexer) getTipsetKeyCidByHeight(ctx context.Context, height abi.ChainEpoch) ([]byte, error) {
+	ts, err := si.cs.GetTipsetByHeight(ctx, height, nil, false)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get tipset by height: %w", err)
+	}
+
+	if ts.Height() != height {
+		// this means that this is a null round
+		return nil, ErrNotFound
+	}
+
+	return toTipsetKeyCidBytes(ts)
+}
+
+// GetEventsForFilter returns matching events for the given filter
+// Returns nil, nil if the filter has no matching events
+// Returns nil, ErrNotFound if the filter has no matching events and the tipset is not indexed
+// Returns nil, err for all other errors
+func (si *SqliteIndexer) GetEventsForFilter(ctx context.Context, f *EventFilter, excludeReverted bool) ([]*CollectedEvent, error) {
+	getEventsFnc := func(stmt *sql.Stmt, values []any) ([]*CollectedEvent, error) {
+		q, err := stmt.QueryContext(ctx, values...)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to query events: %w", err)
+		}
+		defer func() { _ = q.Close() }()
+
+		var ces []*CollectedEvent
+		var currentID int64 = -1
+		var ce *CollectedEvent
+
+		for q.Next() {
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			default:
+			}
+
+			var row struct {
+				id           int64
+				height       uint64
+				tipsetKeyCid []byte
+				emitterID    uint64
+				emitterAddr  []byte
+				eventIndex   int
+				messageCid   []byte
+				messageIndex int
+				reverted     bool
+				flags        []byte
+				key          string
+				codec        uint64
+				value        []byte
+			}
+
+			if err := q.Scan(
+				&row.id,
+				&row.height,
+				&row.tipsetKeyCid,
+				&row.emitterID,
+				&row.emitterAddr,
+				&row.eventIndex,
+				&row.messageCid,
+				&row.messageIndex,
+				&row.reverted,
+				&row.flags,
+				&row.key,
+				&row.codec,
+				&row.value,
+			); err != nil {
+				return nil, xerrors.Errorf("read prefill row: %w", err)
+			}
+
+			if row.id != currentID {
```

Comment:
```suggestion
			// The query will return all entries for all matching events, so we need to keep track
			// of which event we are dealing with and create a new time each time we see a new id
			if row.id != currentID {
```

just had a cognitive hurdle here, comment always helps with that